pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -0,0 +1,220 @@
+		case _, ok := <-doneCh:
+			if !ok {
+				return
+			}
+		case errCh <- err:
+			close(errCh)
+			return
+		}
+	}
+	defer cancel()
+
+	requests := make([]Request, maxMergeRequests+1)
+	for {
+		select {
+		case first := <-tsoRequestCh:
+			pendingTSOReqCount := len(tsoRequestCh) + 1
+			requests[0] = first
+			for i := 1; i < pendingTSOReqCount; i++ {
+				requests[i] = <-tsoRequestCh
+			}
+			done := make(chan struct{})
+			dl := deadline{
+				timer:  time.After(DefaultTSOProxyTimeout),
+				done:   done,
+				cancel: cancel,
+			}
+			select {
+			case tsDeadlineCh <- dl:
+			case <-dispatcherCtx.Done():
+				return
+			}
+			err = s.processRequests(forwardStream, requests[:pendingTSOReqCount], tsoProtoFactory)
+			close(done)
+			if err != nil {
+				log.Error("proxy forward tso error",
+					zap.String("forwarded-host", forwardedHost),
+					errs.ZapError(errs.ErrGRPCSend, err))
+				select {
+				case <-dispatcherCtx.Done():
+					return
+				case _, ok := <-doneCh:
+					if !ok {
+						return
+					}
+				case errCh <- err:
+					close(errCh)
+					return
+				}
+			}
+		case <-dispatcherCtx.Done():
+			return
+		}
+	}
+}
+
+func (s *TSODispatcher) processRequests(forwardStream stream, requests []Request, tsoProtoFactory ProtoFactory) error {
+	// Merge the requests
+	count := uint32(0)
+	for _, request := range requests {
+		count += request.getCount()
+	}
+
+	start := time.Now()
+	resp, err := requests[0].process(forwardStream, count, tsoProtoFactory)
+	if err != nil {
+		return err
+	}
+	s.tsoProxyHandleDuration.Observe(time.Since(start).Seconds())
+	s.tsoProxyBatchSize.Observe(float64(count))
+	// Split the response
+	ts := resp.GetTimestamp()
+	physical, logical, suffixBits := ts.GetPhysical(), ts.GetLogical(), ts.GetSuffixBits()
+	// `logical` is the largest ts's logical part here, we need to do the subtracting before we finish each TSO request.
+	// This is different from the logic of client batch, for example, if we have a largest ts whose logical part is 10,
+	// count is 5, then the splitting results should be 5 and 10.
+	firstLogical := addLogical(logical, -int64(count), suffixBits)
+	return s.finishRequest(requests, physical, firstLogical, suffixBits)
+}
+
+// Because of the suffix, we need to shift the count before we add it to the logical part.
+func addLogical(logical, count int64, suffixBits uint32) int64 {
+	return logical + count<<suffixBits
+}
+
+func (s *TSODispatcher) finishRequest(requests []Request, physical, firstLogical int64, suffixBits uint32) error {
+	countSum := int64(0)
+	for i := 0; i < len(requests); i++ {
+		newCountSum, err := requests[i].postProcess(countSum, physical, firstLogical, suffixBits)
+		if err != nil {
+			return err
+		}
+		countSum = newCountSum
+	}
+	return nil
+}
+
+type deadline struct {
+	timer  <-chan time.Time
+	done   chan struct{}
+	cancel context.CancelFunc
+}
+
+func watchTSDeadline(ctx context.Context, tsDeadlineCh <-chan deadline) {
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
Do we no longer need `defer logutil.LogPanic()`?

Reply:
added back

Reply:
We don't need "defer logutil.LogPanic()". Somehow I added "defer logutil.LogPanic()" in the copy of tso. The original copy used by pd server doesn't have "defer logutil.LogPanic()".